spec/integration/application/agent_spec.rb

```diff
@@ -0,0 +1,233 @@
+        }.to exit_with(0)
+      end
+    end
+
+    it "falls back, recording the first viable server in the report" do
+      server.start_server do |port|
+        Puppet[:server_list] = "puppet.example.com,#{Puppet[:server]}"
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+
+        report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+        expect(report.master_used).to eq("127.0.0.1:#{port}")
+      end
+    end
+
+    it "doesn't write a report if no servers could be contacted" do
+      Puppet[:server_list] = "puppet.example.com"
+
+      expect {
+        agent.command_line.args = args
+        agent.run
+      }.to exit_with(1)
+
+      # I'd expect puppet to update the last run report even if the server_list was
+      # exhausted, but it doesn't work that way currently, see PUP-6708
+      expect(File).to_not be_exist(Puppet[:lastrunreport])
+    end
+
+    it "omits master_used when not using server_list" do
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+      end
+
+      report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+      expect(report.master_used).to be_nil
+    end
+  end
+
+  context 'rich data' do
+    it "applies deferred values" do
+      catalog_handler = -> (req, res) {
+        catalog = compile_to_catalog(<<-MANIFEST, node)
+          notify { 'deferred':
+            message => Deferred('join', [[1,2,3], ':'])
+          }
+        MANIFEST
+
+        res.body = formatter.render(catalog)
+        res['Content-Type'] = formatter.mime
+      }
+
+      server.start_server(mounts: {catalog: catalog_handler}) do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(2)
+      end
+    end
+
+    it "redacts sensitive values" do
+      catalog_handler = -> (req, res) {
+        catalog = compile_to_catalog(<<-MANIFEST, node)
+          notify { 'sensitive':
+            message => Sensitive('supersecret')
+          }
+        MANIFEST
+
+        res.body = formatter.render(catalog)
+        res['Content-Type'] = formatter.mime
+      }
+
+      server.start_server(mounts: {catalog: catalog_handler}) do |port|
+        Puppet[:masterport] = port
+        expect {
+          expect {
+            # don't redirect to logfile so we can assert stdout
+            agent.command_line.args = ['--test']
+            agent.run
+          }.to exit_with(2)
+        }.to output(a_string_matching(
+          /Notice: Sensitive \[value redacted\]/
+        ).and matching(
+          /Notify\[sensitive\]\/message: changed \[redacted\] to \[redacted\]/
+        )).to_stdout
+      end
+    end
+  end
+
+  context 'static catalogs' do
```

Comment:
These tests are based on https://github.com/puppetlabs/puppet/blob/6.14.0/acceptance/tests/direct_puppet/static_catalog_env_control.rb, mainly to verify the http client's `static_file_content` REST API works. Since puppetserver is responsible for creating the static catalog (inlining file metadata(s)), we should leave the beaker tests.